File: Planarian.Web/src/Modules/Caves/Models/CaveSearchVm.ts
import {
  capitalizeFirstLetter,
  nameof,
} from "../../../Shared/Helpers/StringHelpers";
import { SelectListItem } from "../../../Shared/Models/SelectListItem";

export interface CaveSearchVm {
  id: string;
  name: string;
  narrativeSnippet?: string;
  reportedOn: string | null;
  isArchived: boolean;
  depthFeet: number | null;
  lengthFeet: number | null;
  maxPitDepthFeet: number | null;
  numberOfPits: number | null;
  county: SelectListItem<string>;
  countyDisplayId: string;
  countyNumber: number;
  displayId: string;
  primaryEntranceLatitude: number | null;
  primaryEntranceLongitude: number | null;
  primaryEntranceElevationFeet: number | null;
  distanceMiles: number | null;
  archaeologyTags: SelectListItem<string>[];
  biologyTags: SelectListItem<string>[];
  cartographerNameTags: SelectListItem<string>[];
  geologicAgeTags: SelectListItem<string>[];
  geologyTags: SelectListItem<string>[];
  mapStatusTags: SelectListItem<string>[];
  otherTags: SelectListItem<string>[];
  physiographicProvinceTags: SelectListItem<string>[];
  reportedByTags: SelectListItem<string>[];
  isFavorite: boolean;
}

export class CaveSearchSortByConstants {
  static readonly DisplayId = capitalizeFirstLetter(
    nameof<CaveSearchVm>("displayId")
  );
  static readonly Name = capitalizeFirstLetter(nameof<CaveSearchVm>("name"));
  static readonly ReportedOn = capitalizeFirstLetter(
    nameof<CaveSearchVm>("reportedOn")
  );
  static readonly DepthFeet = capitalizeFirstLetter(
    nameof<CaveSearchVm>("depthFeet")
  );
  static readonly LengthFeet = capitalizeFirstLetter(
    nameof<CaveSearchVm>("lengthFeet")
  );
  static readonly MaxPitDepthFeet = capitalizeFirstLetter(
    nameof<CaveSearchVm>("maxPitDepthFeet")
  );
  static readonly NumberOfPits = capitalizeFirstLetter(
    nameof<CaveSearchVm>("numberOfPits")
  );
  static readonly DistanceMiles = capitalizeFirstLetter(
    nameof<CaveSearchVm>("distanceMiles")
  );
}

export const CaveSortMetadataKeys = Object.freeze({
  distanceLocation: "distanceSortLocation",
});

File: Planarian/Planarian/Modules/Caves/Models/CaveSearchVm.cs
using Planarian.Model.Shared;

namespace Planarian.Modules.Caves.Models;

public class CaveSearchVm
{
    public string Id { get; set; }
    public string Name { get; set; }
    
    public string? NarrativeSnippet { get; set; }
    
    public DateTime? ReportedOn { get; set; }
    public bool IsArchived { get; set; }
    public double? DepthFeet { get; set; }
    public double? LengthFeet { get; set; }
    public double? MaxPitDepthFeet { get; set; }
    public int? NumberOfPits { get; set; }
    public SelectListItem<string> County { get; set; }
    public string CountyDisplayId { get; set; } = null!;
    public int CountyNumber { get; set; }
    public string DisplayId { get; set; }
    
    public double? PrimaryEntranceLatitude { get; set; }
    public double? PrimaryEntranceLongitude { get; set; }
    public double? PrimaryEntranceElevationFeet { get; set; }

    public double? DistanceMiles { get; set; }

    public bool IsFavorite { get; set; }

    public IEnumerable<SelectListItem<string>> ArchaeologyTags { get; set; }
    public IEnumerable<SelectListItem<string>> BiologyTags { get; set; }
    public IEnumerable<SelectListItem<string>> CartographerNameTags { get; set; }
    public IEnumerable<SelectListItem<string>> GeologicAgeTags { get; set; }
    public IEnumerable<SelectListItem<string>> GeologyTags { get; set; }
    public IEnumerable<SelectListItem<string>> MapStatusTags { get; set; }
    public IEnumerable<SelectListItem<string>> OtherTags { get; set; }
    public IEnumerable<SelectListItem<string>> PhysiographicProvinceTags { get; set; }
    public IEnumerable<SelectListItem<string>> ReportedByTags { get; set; }
}
